esmvaltool/diag_scripts/ensclus/ensclus.py

```diff
@@ -89,23 +92,29 @@ def main(cfg):
                         numens, cfg['season'], cfg['area'], cfg['extreme'])
 
     # ###################### EOF AND K-MEANS ANALYSES #######################
-    outfiles2 = ens_eof_kmeans(out_dir, name_outputs, numens, cfg['numpcs'],
-                               cfg['perc'], cfg['numclus'])
+    outfiles2 = ens_eof_kmeans(out_dir, name_outputs, numens, numpcs,
+                               perc, cfg['numclus'])
 
     outfiles = outfiles + outfiles2
-    provenance_record = get_provenance_record(
+    nc_provenance_record = get_provenance_record(
         cfg, list(files_dict.values())[0][0], ancestor_files=filenames_cat)
+    fig_provenance_record = get_provenance_record(
+        cfg, list(files_dict.values())[0][0], ancestor_files=outfiles)
 
     # ###################### PLOT AND SAVE FIGURES ##########################
     if write_plots:
-        plot_file = ens_plots(out_dir, cfg['plot_dir'], name_outputs,
+        plotfiles = ens_plots(out_dir, cfg['plot_dir'], name_outputs,
                               cfg['numclus'], 'anomalies',
-                              cfg['output_file_type'])
-        provenance_record['plot_file'] = plot_file
+                              cfg['output_file_type'], cfg['season'],
+                              cfg['area'], cfg['extreme'], max_plot_panels)
+        nc_provenance_record['plot_file'] = plotfiles[0]  # waiting for #801
 
     for file in outfiles:
         with ProvenanceLogger(cfg) as provenance_logger:
-            provenance_logger.log(file, provenance_record)
+            provenance_logger.log(file, nc_provenance_record)
```

Comment:
It would be better to write this as:
```python
with ProvenanceLogger(cfg) as provenance_logger:
    for file in outfiles:
        provenance_logger.log(file, nc_provenance_record)
    for file in plotfiles
        provenance_logger.log(file, fig_provenance_record)
```

Reply:
Yes, this is how I did it originally, but it does not write a correct diagnostic_provenance.yml file!
A small mistery.
 
Anyway, I have run into another error: as you will notice I indicate the diagnostic files as ancestors for the plot files. The problem is that when esmvaltool processes diagnostic_provenance.yml  after running the diagnostic, it starts with attaching provenance to the plots and only later it does so for the diagnostic .nc files. The reason is that it probably does this because it proceeds in alphabetical order of the filenames (the keys in the dict) and actually "\*/plots/\*." comes before "\*/work/\*" 
The result is that while provenance is attached to my plot files, it does not mention the diagnostic files and does not copy their provenance (because they do not have it yet when it tries to do so....)

Reply:
> Yes, this is how I did it originally, but it does not write a correct diagnostic_provenance.yml file!
A small mistery.

I will look into it

> Anyway, I have run into another error: as you will notice I indicate the diagnostic files as ancestors for the plot files. The problem is that when esmvaltool processes diagnostic_provenance.yml after running the diagnostic, it starts with attaching provenance to the plots and only later it does so for the diagnostic .nc files. The reason is that it probably does this because it proceeds in alphabetical order of the filenames (the keys in the dict) and actually "*/plots/*." comes before "*/work/*"
The result is that while provenance is attached to my plot files, it does not mention the diagnostic files and does not copy their provenance (because they do not have it yet when it tries to do so....)

This is because you're trying to use the system for something that it was not designed to do. Strictly speaking I do not think you're actually deriving the plots from the diagnostic netcdf files, so I think it's fine if those are not in the plot provenance.

Reply:
> Strictly speaking I do not think you're actually deriving the plots from the diagnostic netcdf files, so I think it's fine if those are not in the plot provenance.

Well, actually yes, `ens_plots.py` reads the ens_anomalies_*nc file created by `ens_eof_kmeans.py` and uses it to create the plot(s). 
Of course if "plot_file" could be a list as I suggested #801 then for this specific diagnostic all would be fine ;) (I know that I am insisting, sorry, but I really think that this specific example - the same plot split into two - makes a good case)

Reply:
> Yes, this is how I did it originally, but it does not write a correct diagnostic_provenance.yml file!

Can you explain what is wrong exactly? Because I changed it locally and ran it and provenance information does get written for all created files, so that looks fine to me.

Reply:
>    Yes, this is how I did it originally, but it does not write a correct diagnostic_provenance.yml file!

I think I found this issue, fixed in  #996. Can you check if this indeed solves the problem for you?

Regarding the multiple plots per file issue: as discussed via email with @ndrost, just avoid using the `plot_file` attribute for now and log provenance for plots the same way you would do it for other files.

Reply:
I tested #996 but it does not change the stranbge diagnostic_provenance.yml file which I get if I use the order
```
with ProvenanceLogger(cfg) as provenance_logger:
    for file in outfiles:
```
See the two example diagnostic_provenance.yml files which I attach below.

Reply:
> Notice the strange ": &id001" in the "notok" file.

That is simply a YAML anchor and should work fine after #996 is merged.